llvm/lib/IR/RepoDefinition.cpp

```diff
@@ -79,187 +88,241 @@ const Constant *getAliasee(const GlobalAlias *GA) {
+                                                      bool HasRepoDefinitionMD)
+    -> std::tuple<size_t, DigestType> {
+  auto LoopPoint = std::numeric_limits<size_t>::max();
   for (const GlobalObject *const G : Dependencies) {
     const llvm::Function *const Fn = dyn_cast<const llvm::Function>(G);
     // if function will not be inlined and not be discarded if it is not used,
     // skip it.
     if (Fn && Fn->hasFnAttribute(Attribute::NoInline) &&
         !Fn->isDiscardableIfUnused())
       continue;
-    Changed = updateDigestUseDependenciesAndContributions(
-                  G, GOHash, Visited, GOIMap, AccumulateGODigest) ||
-              Changed;
+    size_t GDepth;
+    DigestType GDigest;
+    std::tie(GDepth, GDigest) =
+        updateDigestUseDependenciesAndContributions(G, HasRepoDefinitionMD);
+    // A GO which loops back to itself doesn't count as a loop.
+    if (G != GO)
+      LoopPoint = std::min(LoopPoint, GDepth);
+    GOHash.update(GDigest.Bytes);
   }
-  return Changed;
+  // Encoded the 'End' in the GOhash after accumulating all GO's Dependencies.
+  GOHash.update(static_cast<char>(Tags::End));
+
+  DigestType Digest;
+  GOHash.final(Digest);
+
+  // If LoopPoint is bigger than GODepth, GO doesn't belong to any loops (except
+  // self-loop). Record GO's hash in memoized hashes.
+  if (LoopPoint > GODepth) {
+    GOHashCache[GO] = Digest;
+    LLVM_DEBUG(dbgs() << "Recording result for \"" << GO->getName() << "\"\n");
+    ++NumMemoizedHashes;
+    ++VisitedCachedGOTimes;
+  } else {
+    ++VisitedUnCachedGOTimes;
+  }
+  return std::make_tuple(LoopPoint, Digest);
 }
 
-// Update the GO's hash value by adding the hash of its dependencies and
-// contributions.
-template <typename Function>
-static bool updateDigestUseDependenciesAndContributions(
-    const GlobalObject *GO, MD5 &GOHash, GOStateMap &Visited, GOInfoMap &GOIMap,
-    Function AccumulateGODigest) {
-  if (GO->isDeclaration())
-    return false;
+auto ModuleHashGenerator::updateDigestUseDependenciesAndContributions(
+    const GlobalObject *GO, bool HasRepoDefinitionMD)
+    -> std::tuple<size_t, DigestType> {
+  assert(!GO->isDeclaration() && "Can only be used for global definitions");
+
+  const auto GODepth = Visited.size();
+  LLVM_DEBUG(dbgs() << "Computing hash for \"" << GO->getName() << "\" (#"
+                    << GODepth << ")\n");
+
+  // If the hash has been memoized, we can return the result immediately.
+  bool Hit;
+  DigestType GODigest;
+  std::tie(GODigest, Hit) = getGOHash(GO);
+  if (Hit) {
+    ++VisitedCachedGOTimes;
+    LLVM_DEBUG(dbgs() << "Returning pre-computed hash for \"" << GO->getName()
+                      << "\"\n");
+    return std::make_tuple(GODepth, GODigest);
+  }
 
+  MD5 GOHash = MD5();
   bool Inserted;
   typename GOStateMap::const_iterator StateIt;
-  std::tie(StateIt, Inserted) = Visited.try_emplace(GO, Visited.size());
+  // Have we previously visited this GO on this path? If so, add to the hash a
```

Comment:
Instead of "add to the hash a back-reference to that GO", consider "add a back-reference to that GO to the hash"

Reply:
Also consider reversing the order of this explanation, detailing if we _haven't_ visited it first. To me this feels like the more natural order as NOT having visited it before must happen before having visited it before, and it the more common case. e.g.

`// Record visit to GO and its depth to detect loops in the future. If we have
// previously visited this GO (on this path), add a back-reference to that
// point and return.`

I think that detailing returning the position to the caller would be better just before line 261 (return std::make_tuple(StateIt-<second, Digest) )

Reply:
Thanks, changed.